pkg/cmd/join/exec.go

```diff
@@ -0,0 +1,166 @@
+// Copyright Contributors to the Open Cluster Management project
+package join
+
+import (
+	"fmt"
+
+	"github.com/ghodss/yaml"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapiv1 "k8s.io/client-go/tools/clientcmd/api/v1"
+	"open-cluster-management.io/clusteradm/pkg/cmd/join/scenario"
+	"open-cluster-management.io/clusteradm/pkg/helpers"
+
+	"github.com/spf13/cobra"
+)
+
+func (o *Options) complete(cmd *cobra.Command, args []string) (err error) {
+
+	o.values = Values{
+		ClusterName: o.clusterName,
+		Hub: Hub{
+			ExternalServerURL: o.hubServerExternal,
+			InternalServerURL: o.hubServerInternal,
+		},
+	}
+	if o.values.Hub.InternalServerURL == "" {
+		o.values.Hub.InternalServerURL = o.values.Hub.ExternalServerURL
+	}
+	return nil
+}
+
+func (o *Options) validate() error {
+	if o.token == "" {
+		return fmt.Errorf("token is missing")
+	}
+	if o.values.Hub.ExternalServerURL == "" {
+		return fmt.Errorf("hub-server is misisng")
+	}
+	if o.values.ClusterName == "" {
+		return fmt.Errorf("name is missing")
+	}
+
+	return nil
+}
+
+func (o *Options) run() error {
+	reader := scenario.GetScenarioResourcesReader()
+
+	bootstrapConfigUnSecure := clientcmdapiv1.Config{
+		// Define a cluster stanza based on the bootstrap kubeconfig.
+		Clusters: []clientcmdapiv1.NamedCluster{
+			{
+				Name: "hub",
+				Cluster: clientcmdapiv1.Cluster{
+					Server:                o.hubServerExternal,
+					InsecureSkipTLSVerify: true,
+				},
+			},
+		},
+		// Define auth based on the obtained client cert.
+		AuthInfos: []clientcmdapiv1.NamedAuthInfo{
+			{
+				Name: "bootstrap",
+				AuthInfo: clientcmdapiv1.AuthInfo{
+					Token: string(o.token),
+				},
+			},
+		},
+		// Define a context that connects the auth info and cluster, and set it as the default
+		Contexts: []clientcmdapiv1.NamedContext{
+			{
+				Name: "bootstrap",
+				Context: clientcmdapiv1.Context{
+					Cluster:   "hub",
+					AuthInfo:  "bootstrap",
+					Namespace: "default",
+				},
+			},
+		},
+		CurrentContext: "bootstrap",
+	}
+
+	bootstrapConfigBytesUnSecure, err := yaml.Marshal(bootstrapConfigUnSecure)
+	if err != nil {
+		return err
+	}
+
+	configUnSecure, err := clientcmd.Load(bootstrapConfigBytesUnSecure)
+	if err != nil {
+		return err
+	}
+	restConfigUnSecure, err := clientcmd.NewDefaultClientConfig(*configUnSecure, &clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	clientUnSecure, err := kubernetes.NewForConfig(restConfigUnSecure)
+	if err != nil {
+		return err
+	}
+
+	ca, err := helpers.GetCACert(clientUnSecure)
```

Comment:
you can  get apiserver and ca from here. 

I think the logic would be
1. get apiserver-endpoint in cluster-info at first
2. if it is not found, use apiserver from the --hub-apiserver.

In this way, you do not need another flag --hub-internal-apiserverr.

Reply:
Done... hope it is what you expected